File: sdk/core/ios/Sources/Runtime/AdapterRuntime.swift
import Foundation
import Dispatch

// MARK: - Circuit Breaker

public final class AdapterCircuitBreaker {
    private let failureThreshold: Int
    private let timeWindowMs: Int64
    private let recoveryTimeMs: Int64
    
    private var failures: [Int64] = []
    private var state: State = .closed
    private var openedAt: Int64 = 0
    private let lock = NSLock()
    
    public enum State { case closed, open, halfOpen }
    
    public init(failureThreshold: Int = 3, timeWindowMs: Int64 = 30_000, recoveryTimeMs: Int64 = 15_000) {
        self.failureThreshold = failureThreshold
        self.timeWindowMs = timeWindowMs
        self.recoveryTimeMs = recoveryTimeMs
    }
    
    public func getState() -> State {
        lock.lock()
        defer { lock.unlock() }
        
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        if state == .open && now - openedAt >= recoveryTimeMs {
            state = .halfOpen
        }
        return state
    }
    
    public func recordSuccess() {
        lock.lock()
        defer { lock.unlock() }
        
        if state == .halfOpen {
            state = .closed
            failures.removeAll()
        }
    }
    
    public func recordFailure() {
        lock.lock()
        defer { lock.unlock() }
        
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        failures.append(now)
        
        // Remove old failures
        failures = failures.filter { now - $0 <= timeWindowMs }
        
        if state == .closed && failures.count >= failureThreshold {
            state = .open
            openedAt = now
        } else if state == .halfOpen {
            state = .open
            openedAt = now
        }
    }
    
    public func isOpen() -> Bool {
        return getState() == .open
    }
}

// MARK: - Retry Policy

public struct RetryPolicy {
    public static func shouldRetry(error: AdapterError, attemptCount: Int) -> Bool {
        guard attemptCount < 2 else { return false }
        
        switch error.code {
        case .networkError, .timeout:
            return true
        case .error:
            return error.detail.lowercased().contains("5xx") || 
                   error.detail.lowercased().contains("transient")
        default:
            return false
        }
    }
    
    public static func jitterMs() -> Int64 {
        return Int64.random(in: 10...100)
    }
}

// MARK: - Hedge Manager

public final class HedgeManager {
    private var p95Latencies: [String: MovingPercentile] = [:]
    private let lock = NSLock()
    
    public init() {}
    
    public func recordLatency(key: String, latencyMs: Int64) {
        lock.lock()
        defer { lock.unlock() }
        
        if p95Latencies[key] == nil {
            p95Latencies[key] = MovingPercentile(windowSize: 100)
        }
        p95Latencies[key]?.add(value: latencyMs)
    }
    
    public func getHedgeDelayMs(key: String) -> Int64? {
        lock.lock()
        defer { lock.unlock() }
        
        return p95Latencies[key]?.getPercentile(p: 0.95)
    }
    
    class MovingPercentile {
        private var values: [Int64] = []
        private let windowSize: Int
        
        init(windowSize: Int) {
            self.windowSize = windowSize
        }
        
        func add(value: Int64) {
            values.append(value)
            if values.count > windowSize {
                values.removeFirst()
            }
        }
        
        func getPercentile(p: Double) -> Int64? {
            guard !values.isEmpty else { return nil }
            let sorted = values.sorted()
            let index = min(Int(Double(sorted.count) * p), sorted.count - 1)
            return sorted[index]
        }
    }
}

// MARK: - Timeout Enforcer

public final class AdapterTimeoutEnforcer {
    public init() {}
    
    public func withTimeout<T: Sendable>(timeoutMs: Int, operation: @escaping @Sendable () async throws -> T) async throws -> T {
        return try await withThrowingTaskGroup(of: T.self) { group in
            group.addTask {
                return try await operation()
            }
            
            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(timeoutMs) * 1_000_000)
                throw AdapterError(code: .timeout, detail: "Operation exceeded \(timeoutMs)ms", vendorCode: nil, recoverable: true)
            }
            
            guard let result = try await group.next() else {
                throw AdapterError(code: .error, detail: "No result", vendorCode: nil, recoverable: false)
            }
            
            group.cancelAll()
            return result
        }
    }
}

// MARK: - Adapter Runtime Wrapper

public final class AdapterRuntimeWrapper {
    private let adapter: AdNetworkAdapterV2
    private let partnerId: String
    private var circuitBreakers: [String: AdapterCircuitBreaker] = [:]
    private let hedgeManager = HedgeManager()
    private let timeoutEnforcer = AdapterTimeoutEnforcer()
    private let lock = NSLock()
    
    public init(adapter: AdNetworkAdapterV2, partnerId: String) {
        self.adapter = adapter
        self.partnerId = partnerId
    }
    
    private func getCircuitBreaker(placement: String) -> AdapterCircuitBreaker {
        lock.lock()
        defer { lock.unlock() }
        
        let key = "\(partnerId):\(placement)"
        if circuitBreakers[key] == nil {
            circuitBreakers[key] = AdapterCircuitBreaker()
        }
        return circuitBreakers[key]!
    }
    
    public func loadInterstitialWithEnforcement(
        placement: String,
        meta: RequestMeta,
        timeoutMs: Int
    ) async throws -> LoadResult {
        let cb = getCircuitBreaker(placement: placement)
        
        // Fast-fail if circuit open
        guard !cb.isOpen() else {
            throw AdapterError(code: .circuitOpen, detail: "Circuit breaker open for \(partnerId):\(placement)", vendorCode: nil, recoverable: true)
        }
        
        let startTime = Date()
        var attemptCount = 0
        var lastError: AdapterError?
        
        while attemptCount < 2 {
            attemptCount += 1
            
            do {
                let result = try await timeoutEnforcer.withTimeout(timeoutMs: timeoutMs) {
                    return await Task.detached(priority: .userInitiated) {
                        return self.adapter.loadInterstitial(placementId: placement, meta: meta, timeoutMs: timeoutMs)
                    }.value
                }
                
                let latency = Int64(Date().timeIntervalSince(startTime) * 1000)
                hedgeManager.recordLatency(key: "\(partnerId):\(placement)", latencyMs: latency)
                cb.recordSuccess()
                return result
                
            } catch let error as AdapterError {
                lastError = error
                
                guard RetryPolicy.shouldRetry(error: error, attemptCount: attemptCount) else {
                    cb.recordFailure()
                    throw error
                }
                
                try await Task.sleep(nanoseconds: UInt64(RetryPolicy.jitterMs()) * 1_000_000)
                
            } catch {
                let wrappedError = AdapterError(code: .error, detail: error.localizedDescription, vendorCode: nil, recoverable: true)
                lastError = wrappedError
                
                guard RetryPolicy.shouldRetry(error: wrappedError, attemptCount: attemptCount) else {
                    cb.recordFailure()
                    throw wrappedError
                }
                
                try await Task.sleep(nanoseconds: UInt64(RetryPolicy.jitterMs()) * 1_000_000)
            }
        }
        
        cb.recordFailure()
        throw lastError ?? AdapterError(code: .error, detail: "Unknown failure after retries", vendorCode: nil, recoverable: false)
    }
    
    public func loadWithHedging(
        placement: String,
        meta: RequestMeta,
        timeoutMs: Int
    ) async throws -> LoadResult {
        guard let hedgeDelay = hedgeManager.getHedgeDelayMs(key: "\(partnerId):\(placement)"),
              hedgeDelay < timeoutMs else {
            return try await loadInterstitialWithEnforcement(placement: placement, meta: meta, timeoutMs: timeoutMs)
        }
        
        return try await withThrowingTaskGroup(of: LoadResult.self) { group in
            // Primary request
            group.addTask {
                return try await self.loadInterstitialWithEnforcement(placement: placement, meta: meta, timeoutMs: timeoutMs)
            }
            
            // Hedged request
            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(hedgeDelay) * 1_000_000)
                return try await self.loadInterstitialWithEnforcement(placement: placement, meta: meta, timeoutMs: timeoutMs - Int(hedgeDelay))
            }
            
            guard let result = try await group.next() else {
                throw AdapterError(code: .error, detail: "No hedge result", vendorCode: nil, recoverable: false)
            }
            
            group.cancelAll()
            return result
        }
    }
    
    @MainActor
    public func showInterstitialOnMain(handle: AdHandle, viewController: AnyObject, callbacks: ShowCallbacks) {
        adapter.showInterstitial(handle: handle, viewController: viewController, callbacks: callbacks)
    }
}

// MARK: - Thread Guard

public struct ThreadGuard {
    public static func assertMainThread(operation: String) {
        assert(Thread.isMainThread, "\(operation) must be called from main thread")
    }
    
    public static func assertNotMainThread(operation: String) {
        assert(!Thread.isMainThread, "\(operation) must not be called from main thread")
    }
}

File: sdk/core/android/src/main/kotlin/runtime/AdapterRuntime.kt
package com.rivalapexmediation.sdk.runtime

import android.os.Handler
import android.os.Looper
import com.rivalapexmediation.sdk.contract.*
import kotlinx.coroutines.*
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import kotlin.random.Random

/**
 * Enforces runtime rules per Adapters.md section 2:
 * - Threading & latency budgets
 * - Circuit breaker
 * - Retry with jitter
 * - Hedging
 */

// MARK: - Circuit Breaker

class CircuitBreaker(
    private val failureThreshold: Int = 3,
    private val timeWindowMs: Long = 30_000,
    private val recoveryTimeMs: Long = 15_000
) {
    private val failures = ConcurrentLinkedQueue<Long>()
    @Volatile private var state = State.CLOSED
    @Volatile private var openedAt: Long = 0
    
    enum class State { CLOSED, OPEN, HALF_OPEN }
    
    fun getState(): State {
        if (state == State.OPEN && System.currentTimeMillis() - openedAt >= recoveryTimeMs) {
            synchronized(this) {
                if (state == State.OPEN && System.currentTimeMillis() - openedAt >= recoveryTimeMs) {
                    state = State.HALF_OPEN
                }
            }
        }
        return state
    }
    
    fun recordSuccess() {
        synchronized(this) {
            if (state == State.HALF_OPEN) {
                state = State.CLOSED
                failures.clear()
            }
        }
    }
    
    fun recordFailure() {
        val now = System.currentTimeMillis()
        failures.add(now)
        
        // Remove old failures outside window
        while (failures.peek()?.let { now - it > timeWindowMs } == true) {
            failures.poll()
        }
        
        synchronized(this) {
            if (state == State.CLOSED && failures.size >= failureThreshold) {
                state = State.OPEN
                openedAt = now
            } else if (state == State.HALF_OPEN) {
                state = State.OPEN
                openedAt = now
            }
        }
    }
    
    fun isOpen(): Boolean = getState() == State.OPEN
}

// MARK: - Timeout Enforcement

class TimeoutEnforcer(private val scope: CoroutineScope) {
    
    suspend fun <T> withTimeout(timeoutMs: Int, block: suspend () -> T): Result<T> {
        return try {
            withTimeout(timeoutMs.toLong()) {
                Result.success(block())
            }
        } catch (e: TimeoutCancellationException) {
            Result.failure(AdapterError.Recoverable(ErrorCode.TIMEOUT, "Operation exceeded ${timeoutMs}ms"))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

// MARK: - Retry Logic

object RetryPolicy {
    fun shouldRetry(error: AdapterError, attemptCount: Int): Boolean {
        if (attemptCount >= 2) return false // Max 1 retry = 2 total attempts
        
        return when (error.code) {
            ErrorCode.NETWORK_ERROR, ErrorCode.TIMEOUT -> true
            ErrorCode.ERROR -> error.detail.contains("5xx", ignoreCase = true) || 
                              error.detail.contains("transient", ignoreCase = true)
            else -> false
        }
    }
    
    fun jitterMs(): Long = Random.nextLong(10, 101) // 10-100ms jitter
}

// MARK: - Hedging

class HedgeManager {
    private val p95Latencies = ConcurrentHashMap<String, MovingPercentile>()
    
    fun recordLatency(key: String, latencyMs: Long) {
        p95Latencies.computeIfAbsent(key) { MovingPercentile(windowSize = 100) }
            .add(latencyMs)
    }
    
    fun getHedgeDelayMs(key: String): Long? {
        val percentile = p95Latencies[key] ?: return null
        return percentile.getPercentile(0.95)?.toLong()
    }
    
    class MovingPercentile(private val windowSize: Int) {
        private val values = ConcurrentLinkedQueue<Long>()
        
        fun add(value: Long) {
            values.add(value)
            while (values.size > windowSize) {
                values.poll()
            }
        }
        
        fun getPercentile(p: Double): Double? {
            if (values.isEmpty()) return null
            val sorted = values.sorted()
            val index = (sorted.size * p).toInt().coerceIn(0, sorted.size - 1)
            return sorted[index].toDouble()
        }
    }
}

// MARK: - Adapter Runtime Wrapper

class AdapterRuntimeWrapper(
    private val adapter: AdNetworkAdapterV2,
    private val partnerId: String,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
) {
    private val circuitBreakers = ConcurrentHashMap<String, CircuitBreaker>()
    private val hedgeManager = HedgeManager()
    private val timeoutEnforcer = TimeoutEnforcer(scope)
    private val mainHandler = Handler(Looper.getMainLooper())
    
    private fun getCircuitBreaker(placement: String): CircuitBreaker {
        return circuitBreakers.computeIfAbsent("$partnerId:$placement") { CircuitBreaker() }
    }
    
    /**
     * Load with full runtime enforcement: timeout, retry, circuit breaker, hedging
     */
    suspend fun loadInterstitialWithEnforcement(
        placement: String,
        meta: RequestMeta,
        timeoutMs: Int
    ): LoadResult {
        val cb = getCircuitBreaker(placement)
        
        // Fast-fail if circuit is open
        if (cb.isOpen()) {
            throw AdapterError.Fatal(ErrorCode.CIRCUIT_OPEN, "Circuit breaker open for $partnerId:$placement")
        }
        
        val startTime = System.currentTimeMillis()
        var attemptCount = 0
        var lastError: AdapterError? = null
        
        // Primary attempt with retry
        while (attemptCount < 2) {
            attemptCount++
            
            try {
                val result = timeoutEnforcer.withTimeout(timeoutMs) {
                    // Ensure no blocking on main thread
                    withContext(Dispatchers.IO) {
                        adapter.loadInterstitial(placement, meta, timeoutMs)
                    }
                }
                
                result.onSuccess { loadResult ->
                    val latency = System.currentTimeMillis() - startTime
                    hedgeManager.recordLatency("$partnerId:$placement", latency)
                    cb.recordSuccess()
                    return loadResult
                }
                
                result.onFailure { throwable ->
                    val error = when (throwable) {
                        is AdapterError -> throwable
                        else -> AdapterError.Recoverable(ErrorCode.ERROR, throwable.message ?: "Unknown error")
                    }
                    lastError = error
                    
                    if (!RetryPolicy.shouldRetry(error, attemptCount)) {
                        throw error
                    }
                    
                    delay(RetryPolicy.jitterMs())
                }
            } catch (e: AdapterError) {
                lastError = e
                if (!RetryPolicy.shouldRetry(e, attemptCount)) {
                    cb.recordFailure()
                    throw e
                }
                delay(RetryPolicy.jitterMs())
            }
        }
        
        cb.recordFailure()
        throw lastError ?: AdapterError.Fatal(ErrorCode.ERROR, "Unknown failure after retries")
    }
    
    /**
     * Load with hedging: start second request at p95 latency
     */
    suspend fun loadWithHedging(
        placement: String,
        meta: RequestMeta,
        timeoutMs: Int
    ): LoadResult = coroutineScope {
        val hedgeDelay = hedgeManager.getHedgeDelayMs("$partnerId:$placement")
        
        if (hedgeDelay == null || hedgeDelay >= timeoutMs) {
            // No hedging data or would exceed timeout
            return@coroutineScope loadInterstitialWithEnforcement(placement, meta, timeoutMs)
        }
        
        // Start primary
        val primary = async { loadInterstitialWithEnforcement(placement, meta, timeoutMs) }
        
        // Start hedge after delay
        val hedge = async {
            delay(hedgeDelay)
            loadInterstitialWithEnforcement(placement, meta, timeoutMs - hedgeDelay.toInt())
        }
        
        // Return first success (race)
        try {
            withTimeout(timeoutMs.toLong()) {
                val primaryResult = primary.await()
                hedge.cancel()
                primaryResult
            }
        } catch (e: TimeoutCancellationException) {
            hedge.await()
        }
    }
    
    /**
     * Show on main thread with guards
     */
    fun showInterstitialOnMain(handle: AdHandle, viewContext: Any, callbacks: ShowCallbacks) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            adapter.showInterstitial(handle, viewContext, callbacks)
        } else {
            mainHandler.post {
                adapter.showInterstitial(handle, viewContext, callbacks)
            }
        }
    }

    fun showRewardedOnMain(handle: AdHandle, viewContext: Any, callbacks: RewardedCallbacks) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            adapter.showRewarded(handle, viewContext, callbacks)
        } else {
            mainHandler.post {
                adapter.showRewarded(handle, viewContext, callbacks)
            }
        }
    }
}

// MARK: - Thread Guard (Debug)

object ThreadGuard {
    fun assertMainThread(operation: String) {
        if (BuildConfig.DEBUG && Looper.myLooper() != Looper.getMainLooper()) {
            throw IllegalStateException("$operation must be called from main thread")
        }
    }
    
    fun assertNotMainThread(operation: String) {
        if (BuildConfig.DEBUG && Looper.myLooper() == Looper.getMainLooper()) {
            throw IllegalStateException("$operation must not be called from main thread")
        }
    }
}

// Stub for BuildConfig
private object BuildConfig {
    const val DEBUG = true
}
